Thinktecture.Relay.Server/Communication/RabbitMq/RabbitMqMessageDispatcher.cs

```diff
@@ -17,123 +16,107 @@ namespace Thinktecture.Relay.Server.Communication.RabbitMq
-						if (!noAck)
-						{
-							_model.BasicAck(args.DeliveryTag, false);
-						}
-					}
+					return;
 				}
 
-				consumer.Received += OnReceived;
-
-				return new DelegatingDisposable(_logger, () =>
+				switch (request.AcknowledgmentMode)
 				{
-					_logger?.Debug("Disposing request consumer. link-id={LinkId}, connection-id={ConnectionId}", linkId, connectionId);
-					consumer.Received -= OnReceived;
-					_model.BasicCancel(consumerTag);
-				});
+					case AcknowledgmentMode.Auto:
+						_model.BasicAck(deliveryTag, false);
+						_logger?.Debug("Request was automatically acknowledged. request-id={RequestId}", request.RequestId);
+						break;
+
+					case AcknowledgmentMode.Default:
+					case AcknowledgmentMode.Manual:
+						request.AcknowledgeId = deliveryTag.ToString();
+						request.AcknowledgeOriginId = _originId;
+						_logger?.Verbose("Request acknowledge id was set. request-id={RequestId}, acknowledge-id={AcknowledgeId}", request.RequestId, request.AcknowledgeId);
+						break;
+				}
 			});
 		}
 
-		public IObservable<IOnPremiseConnectorResponse> OnResponseReceived(Guid originId)
+		public IObservable<IOnPremiseConnectorResponse> OnResponseReceived()
 		{
-			return Observable.Create<IOnPremiseConnectorResponse>(observer =>
-			{
-				var queueName = "Response " + originId;
+			return CreateConsumerObservable<OnPremiseConnectorResponse>($"{_RESPONSE_QUEUE_PREFIX}{_originId}");
+		}
 
+		public IObservable<string> OnAcknowledgeReceived()
+		{
+			return CreateConsumerObservable<string>($"{_ACKNOWLEDGE_QUEUE_PREFIX}{_originId}");
+		}
+
+		private IObservable<T> CreateConsumerObservable<T>(string queueName, bool autoAck = true, Action<T, ulong> callback = null)
+		{
+			return Observable.Create<T>(observer =>
+			{
 				DeclareQueue(queueName);
-				_model.QueueBind(queueName, _EXCHANGE_NAME, originId.ToString());
+				_model.QueueBind(queueName, _EXCHANGE_NAME, queueName);
 
-				_logger?.Debug("Creating response consumer");
+				_logger?.Debug("Creating consumer. queue-name={QueueName}", queueName);
 
 				var consumer = new EventingBasicConsumer(_model);
-				var consumerTag = _model.BasicConsume(queueName, true, consumer);
+				var consumerTag = _model.BasicConsume(queueName, autoAck, consumer);
 
 				void OnReceived(object sender, BasicDeliverEventArgs args)
 				{
 					try
 					{
 						var json = _encoding.GetString(args.Body);
-						var request = JsonConvert.DeserializeObject<OnPremiseConnectorResponse>(json);
+						var message = JsonConvert.DeserializeObject<T>(json);
+
+						callback?.Invoke(message, args.DeliveryTag);
 
-						observer.OnNext(request);
+						observer.OnNext(message);
 					}
 					catch (Exception ex)
 					{
-						_logger?.Error(ex, "Error during reception of an request via RabbitMQ");
+						_logger?.Error(ex, "Error during receiving a message via RabbitMQ");
+
+						if (!autoAck)
+						{
+							_model.BasicAck(args.DeliveryTag, false);
```

Comment:
Maybe put that in an additional try/catch? If reading a message from the queue throws an exception, it could be likely that ACKing it throws, too.

Reply:
Results to nothing, as the model would anyway dead. This should be handled in a way bigger scope (and a separate PR, of course).